### PR TITLE
Add dartdoc to basic.dart

### PR DESCRIPTION
**Pull request by abarth**
This patch starts working through the missing dartdocs in widgets.dart.

### CODE REVIEW

**Comment by abarth** [comment]
@Hixie 

I started using `@required` in this patch in a couple cases:

 * Argument that can't be null and don't have default values.
 * The `child` argument to widget that don't make sense without a child (e.g., ParentDataWidgets).

**Comment by abarth** [comment]
`548 public members lack documentation`

**Comment by Hixie** [comment]
Anything that isn't a RenderObjectWidget should have a required child.

**Comment by abarth** [comment]
> Anything that isn't a RenderObjectWidget should have a required child.

We often use Container without a child.

File: packages/flutter/lib/src/material/dialog.dart [hunk 1]
```diff
@@ -37,15 +37,15 @@ class Dialog extends StatelessWidget {
   /// of the dialog.
   final Widget title;
 
-  // Padding around the title; uses material design default if none is supplied
-  // If there is no title, no padding will be provided
+  /// Padding around the title; uses material design default if none is supplied
+  /// If there is no title, no padding will be provided
```
Comment (Hixie) [inline packages/flutter/lib/src/material/dialog.dart]: period

File: packages/flutter/lib/src/material/dialog.dart [hunk 2]
```diff
@@ -37,15 +37,15 @@ class Dialog extends StatelessWidget {
...
   final EdgeInsets titlePadding;
 
   /// The (optional) content of the dialog is displayed in the center of the
   /// dialog in a lighter font.
   final Widget content;
 
-  // Padding around the content; uses material design default if none is supplied
+  /// Padding around the content; uses material design default if none is supplied
```
Comment (Hixie) [inline packages/flutter/lib/src/material/dialog.dart]: period

File: packages/flutter/lib/src/widgets/app.dart [hunk 3]
```diff
@@ -96,12 +100,14 @@ class WidgetsApp extends StatefulWidget {
   /// The observer for the Navigator created for this app.
   final NavigatorObserver navigatorObserver;
 
+  /// If true, forces the performance overlay to be visible in all instances.
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/app.dart]: this would be a good place to point to the observatory extension that uses this

File: packages/flutter/lib/src/widgets/app.dart [hunk 4]
```diff
@@ -96,12 +100,14 @@ class WidgetsApp extends StatefulWidget {
...
   static bool showPerformanceOverlayOverride = false;
 
   @override
   WidgetsAppState<WidgetsApp> createState() => new WidgetsAppState<WidgetsApp>();
 }
 
+/// State for subclasses of [WidgetsApp].
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/app.dart]: this can be private now, i think

Comment (Hixie) [inline packages/flutter/lib/src/widgets/app.dart]: and doesn't need to be generic

File: packages/flutter/lib/src/widgets/asset_vendor.dart [hunk 5]
```diff
@@ -197,21 +201,27 @@ class _ResolutionAwareAssetResolver extends _VariantAssetResolver {
...
+  /// The bundle from which to load the assets.
   final AssetBundle bundle;
 
+  /// If non-null, the device pixel ratio to assume when selecting assets.
   final double devicePixelRatio;
 
   /// The widget below this widget in the tree.
   final Widget child;
 
+  /// The function to use for decoding images.
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/asset_vendor.dart]: please keep the members in the same order as the constructor arguments

File: packages/flutter/lib/src/widgets/auto_layout.dart [hunk 6]
```diff
@@ -28,10 +34,28 @@ class AutoLayout extends MultiChildRenderObjectWidget {
   }
 }
 
+/// A widget that provides constraints for a child of an [AutoLayout] widget.
+///
+/// This widget must be a descendant of an [AutoLayout], and the path from this
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/auto_layout.dart]: not clear what "This widget" means here'

Comment (Hixie) [inline packages/flutter/lib/src/widgets/auto_layout.dart]: Maybe the problem is with the word "child" on the first line?

File: packages/flutter/lib/src/widgets/auto_layout.dart [hunk 7]
```diff
@@ -28,10 +34,28 @@ class AutoLayout extends MultiChildRenderObjectWidget {
...
+/// widget to its enclosing [AutoLayout] must contain only [StatelessWidget]s or
+/// [StatefulWidget]s (not other kinds of widgets, like [RenderObjectWidget]s).
 class AutoLayoutChild extends ParentDataWidget<AutoLayout> {
-  AutoLayoutChild({ AutoLayoutRect rect, Widget child })
-    : rect = rect, super(key: new ObjectKey(rect), child: child);
+  /// Creates a widget that provides constraints for a child of an [AutoLayout] widget.
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/auto_layout.dart]: if child is wrong above, it's probably also wrong here

File: packages/flutter/lib/src/widgets/auto_layout.dart [hunk 8]
```diff
@@ -28,10 +34,28 @@ class AutoLayout extends MultiChildRenderObjectWidget {
...
+  ///
+  /// The [rect] argument must be unique among children of a given [AutoLayout]
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/auto_layout.dart]: unique in what sense?

File: packages/flutter/lib/src/widgets/auto_layout.dart [hunk 9]
```diff
@@ -28,10 +34,28 @@ class AutoLayout extends MultiChildRenderObjectWidget {
...
+  /// widget.
+  AutoLayoutChild({
+    AutoLayoutRect rect,
+    @required Widget child
+  }) : rect = rect,
+       super(key: rect != null ? new ObjectKey(rect) : null, child: child);
 
+  /// The constraints to use for this child.
+  ///
+  /// The [rect] object must be unique among children of a given [AutoLayout]
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/auto_layout.dart]: ditto regarding unique

File: packages/flutter/lib/src/widgets/basic.dart [hunk 10]
```diff
@@ -174,14 +181,19 @@ class BackdropFilter extends SingleChildRenderObjectWidget {
   }
 }
 
-/// Paints a [Decoration] either before or after its child paints.
-/// Container insets its child by the widths of the borders, this Widget does not.
+/// A widget that paints a [Decoration] either before or after its child paints.
+///
+/// Container insets its child by the widths of the borders, this widget does not.
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/basic.dart]: comma should be semicolon here

Comment (Hixie) [inline packages/flutter/lib/src/widgets/basic.dart]: [Container]

File: packages/flutter/lib/src/widgets/basic.dart [hunk 11]
```diff
@@ -1224,6 +1286,9 @@ class Baseline extends SingleChildRenderObjectWidget {
 /// Viewport is the core scrolling primitive in the system, but it can be used
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/basic.dart]: these docs should talk about the importance of [mainAxis].

File: packages/flutter/lib/src/widgets/basic.dart [hunk 12]
```diff
@@ -1306,21 +1387,21 @@ class Container extends StatelessWidget {
   /// If null, the container will expand to fill all available space in its parent.
   final Widget child;
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/basic.dart]: should we have a `static final Container empty = new Container(width: 0.0, height: 0.0)` ?

File: packages/flutter/lib/src/widgets/basic.dart [hunk 13]
```diff
@@ -1388,17 +1469,24 @@ class Container extends StatelessWidget {
...
+/// A widget that uses the block layout algorithm for its children.
 ///
 /// This widget is rarely used directly. Instead, consider using [Block], which
 /// combines the block layout algorithm with scrolling behavior.
 ///
 /// For details about the block layout algorithm, see [RenderBlockBase].
+///
+/// See also:
+///
+///  * [Block] (which combines block layout with scrolling)
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/basic.dart]: I know I started this trend, but I think we should move from `* foo (which blah)` to `* foo, which blah` (or `*foo, blah` in some cases). The parentheses don't work in all the cases where I want to use this pattern. I can do that in a later pass though.

File: packages/flutter/lib/src/widgets/basic.dart [hunk 14]
```diff
@@ -1634,8 +1722,18 @@ class Positioned extends ParentDataWidget<Stack> {
   }
 }
 
-abstract class GridRenderObjectWidgetBase extends MultiChildRenderObjectWidget {
-  GridRenderObjectWidgetBase({
+/// A [MultiChildRenderObjectWidget] for use with [RenderGrid].
+///
+/// Typically not used directly. Instead, consider using one of its subclasses.
+///
+/// See also:
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/basic.dart]: maybe instead of "See also", just say "...subclasses, such as:" ?

File: packages/flutter/lib/src/widgets/basic.dart [hunk 15]
```diff
@@ -1656,12 +1754,21 @@ abstract class GridRenderObjectWidgetBase extends MultiChildRenderObjectWidget {
...
 
-/// Uses the grid layout algorithm for its children.
+/// A widget that creates a grid using a custom [delegate].
 ///
 /// For details about the grid layout algorithm, see [RenderGrid].
-class CustomGrid extends GridRenderObjectWidgetBase {
-  CustomGrid({ Key key, List<Widget> children: _emptyWidgetList, this.delegate })
-    : super(key: key, children: children) {
+///
+/// To pass data to your delegates [GridDelegate.getChildPlacement] function,
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/basic.dart]: delegate's

File: packages/flutter/lib/src/widgets/basic.dart [hunk 16]
```diff
@@ -1656,12 +1754,21 @@ abstract class GridRenderObjectWidgetBase extends MultiChildRenderObjectWidget {
...
+/// see [GridPlacementData].
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/basic.dart]: "to do x, see y" doesn't make sense. You don't just see y to do x. you see y to learn how to do x. :-)

File: packages/flutter/lib/src/widgets/auto_layout.dart (hunk 6 above)
Comment (abarth) [inline packages/flutter/lib/src/widgets/auto_layout.dart]: Fixed by being more explicit.

File: packages/flutter/lib/src/widgets/basic.dart [hunk 17]
```diff
@@ -1758,10 +1873,16 @@ class MaxTileWidthGrid extends GridRenderObjectWidgetBase {
   }
 }
 
-/// Supplies per-child data to the grid's [GridDelegate].
-class GridPlacementData<DataType, WidgetType extends RenderObjectWidget> extends ParentDataWidget<WidgetType> {
-  GridPlacementData({ Key key, this.placementData, Widget child })
-    : super(key: key, child: child);
+/// A widget that controls the placement of a child in a grid.
+class GridPlacementData<DataType> extends ParentDataWidget<GridRenderObjectWidget> {
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/basic.dart]: If CustomGrid is going to point here for details, we should add a lot more detail here.

Comment (Hixie) [inline packages/flutter/lib/src/widgets/basic.dart]: Based on these docs, and the code around them, and searching for "GridPlacementData" in our repo, I have no idea how this is supposed to be used.

File: packages/flutter/lib/src/widgets/basic.dart (hunk 12 above)
Comment (abarth) [inline packages/flutter/lib/src/widgets/basic.dart]: Any `empty` widget would make more sense as a SizedBox...

File: packages/flutter/lib/src/widgets/basic.dart (hunk 14 above)
Comment (abarth) [inline packages/flutter/lib/src/widgets/basic.dart]: done.

File: packages/flutter/lib/src/widgets/basic.dart (hunk 16 above)
Comment (abarth) [inline packages/flutter/lib/src/widgets/basic.dart]: changed to "consider using"

**Comment by Hixie** [comment]
LGTM

This is great.

File: packages/flutter/lib/src/widgets/basic.dart (hunk 17 above)
Comment (abarth) [inline packages/flutter/lib/src/widgets/basic.dart]: I wrote moar docs

**Comment by Hixie** [comment]
>> Anything that isn't a RenderObjectWidget should have a required child.
> We often use Container without a child.

Let me try again.

An argument X of type `Widget` in a subclass W of `Widget` is not required if:

 * W is a `RenderObjectWidget`
 * W's build function does not use X if X is null.
 * W's build function uses X as an argument to a constructor of another `Widget` subclass, and that argument is itself not required.

In all other cases, an argument of type `Widget` in a subclass `Widget` should be required.

**Comment by abarth** [comment]
That sounds roughly right but something I'd probably need a tool to help me with.  :)
